packages/offix-server-conflicts/src/states/VersionedObjectState.ts

```diff
@@ -0,0 +1,52 @@
+import { ObjectState } from "../api/ObjectState";
+import { ObjectStateData } from "../api/ObjectStateData";
+import { ObjectConflictError } from "../api/ObjectConflictError";
+import { conflictHandler } from "..";
+
+/**
+ * Object state manager using a version field
+ * Detects conflicts and allows moving to next state using the version field of the object
+ *
+ * VersionedObjectState requires GraphQL types to contain version field.
+ * For example:
+ *
+ * type User {
+ *   id: ID!
+ *   version: String
+ * }
+ */
+export class VersionedObjectState implements ObjectState {
+
+  public checkForConflict(serverState: ObjectStateData, clientState: ObjectStateData) {
+    if (serverState.version && clientState.version) {
+      if (serverState.version !== clientState.version) {
+        const filteredServerState: any = {};
+        for (const key of Object.keys(clientState)) {
+          filteredServerState[key] = serverState[key];
+        }
+        throw new ObjectConflictError({
```

Comment:
The more I think about this, the stronger I feel it is not the best approach. The function is called `checkForConflict` and if one is found, it instantly throws. I feel like this is a highly opinionated approach and is probably a strong anti-pattern.

I think it will come back to bite us. The fact we are throwing an error like this is based off our current understanding of Apollo Server error handling, which has changed in the past and may change again. In a way, this is creating a loose dependency on Apollo Server internals and the user does not even realise it. Also, just in general it's really weird that the user would explicitly ask 'Is there a conflict?' and we say 'Yes, here is the info' via throwing an error.

As an end user, I would not expect an error to be thrown by this function. I think we should instead return the error object and tell the user to either return/throw it themselves. My understanding is you don't even need to throw because Apollo Server checks the return value and will do the same error handling there too.

This would mean the user would write some code like this.

```js
const conflict = conflictHandler.checkForConflict(serverData,clientData)

if (conflict) {
  // as an end user/developer I can perform any other tasks I need as part of conflicts
  // logging
  // update metrics/monitoring
  // send email to boss :)

  throw conflict // or just return it
}

// rest of resolver
```

I think this approach would be much more beneficial.

* The code written by the end user is much more explicit and readable. The intention is clear.
* The user can have additional steps for dealing with conflicts (logging, metrics, send email to the boss) without having to write unnecessary `try/catch` blocks everywhere in their resolvers.
* We do not make any assumptions in our internal code about how errors are handled by the GraphQL environment. apollo-server vs graphql-js vs graphql-express etc.

@wtrocki @StephenCoady what do you think?

Reply:
Brilliant idea. Going to apply this now